inc/t_cose/t_cose_encrypt_dec.h

```diff
@@ -0,0 +1,200 @@
+ * encryption with a particular set of algorithms. At this moment, only
+ * the integration with Mbed TLS (and more specifically the PSA Crypto
+ * API) is supported.
+ *
+ * See t_cose_common.h for preprocessor defines to reduce object code
+ * and stack use by disabling features.
+ *
+ * A recipient of a COSE_Encrypt0 or a COSE_Encrypt message knows what key
+ * distribution mechanism is utilized in a specific context since the keys
+ * have to be available accordingly.
+ *
+ * 1. Import a key, for example with psa_import_key().
+ *    This key may be the pre-agreed CEK (in case of direct key distribution)
+ *    or the private key corresponding to the public key used by the sender
+ *    (in case of key agreement with HPKE).
+ * 2. Initialize the decrypt context with t_cose_encrypt_dec_init() for use
+ *    with a selected key distribution mechanism.
+ * 3. Use t_cose_encrypt_dec_set_private_key() to associate the previously
+ *    imported key with the decryption structure.
+ * 4. Call t_cose_encrypt_dec() to decrypt the ciphertext.
+ */
+
+/**
+ * Support key distribution types for use with t_cose_encrypt_dec_init().
+ */
+#define T_COSE_KEY_DISTRIBUTION_DIRECT 0x00000001
+#define T_COSE_KEY_DISTRIBUTION_HPKE   0x00000002
+
+/**
+ * Context for use with decryption.
+ */
+struct t_cose_encrypt_dec_ctx {
+    /* Private data structure */
+    int32_t               key_distribution;
+    uint32_t              option_flags;
+    struct q_useful_buf_c kid;
+    struct t_cose_key     recipient_key;
+};
+
+/**
+ * \brief Initialize context for \c COSE_Encrypt and \c COSE_Encrypt0
+ * decryption.
+ *
+ * \param[in]      context           The context to initialize.
+ * \param[in]      option_flags      Options controlling the encryption.
+ *                                   Currently none.
+ * \param[in]      key_distribution  Key distribution setting
+ */
+static void
+t_cose_encrypt_dec_init(struct t_cose_encrypt_dec_ctx *context,
+                        uint32_t                       option_flags,
+                        uint32_t                       key_distribution);
+
+
+/**
+ * \brief Set private key for decryption of \c COSE_Encrypt and
+ * \c COSE_Encrypt0.
+ *
+ * \param[in] context       The t_cose_encrypt_dec_ctx context.
+ * \param[in] key           The private key.
+ * \param[in] kid           The key identifier.
+ *
+ * Important: The key distribution mechanism determines what type of key
+ * is provided. When direct key management is used then the key parameter
+ * contains a symmetric key (and the \c COSE_Encrypt0 structure is assumed).
+ * For use with HPKE an asymmetric private key has to be provided and the
+ * \c COSE_Encrypt structure is assumed.
+ */
+static void
+t_cose_encrypt_dec_set_private_key(struct t_cose_encrypt_dec_ctx *context,
+                                   struct t_cose_key              key,
+                                   struct q_useful_buf_c          kid);
+
+
```

Comment:
An area of future work is something to be able to get the headers without decrypting. This is necessary to get kid and such. It is probably part of work for generalizing recipient implementation. Hopefully it doesn't affect any of the other interfaces here. Doesn't seem like it to me.